ambry-server/src/main/java/com/github/ambry/server/AmbryServer.java

```diff
@@ -263,16 +255,44 @@ public void startup() throws InstantiationException {
          * having a separate CloudtoStorageManager for cloud offers the flexibility to override methods and add more
          * suited to cloud.
          */
-        networkClientFactory = new RecoveryNetworkClientFactory(properties, registry, clusterMap, storageManager,
-            accountService);
-        vcrClusterSpectator = null; // Server does not talk to vcr during recovery
-        _recoveryManager =
-            new RecoveryManager(replicationConfig, clusterMapConfig, storeConfig, storageManager,
-                storeKeyFactory, clusterMap, scheduler, nodeId, networkClientFactory, registry, notificationSystem,
-                storeKeyConverterFactory, serverConfig.serverMessageTransformer, null,
-                clusterParticipants.get(0));
-        _recoveryManager.start();
-      } else {
+        networkClientFactory =
+            new RecoveryNetworkClientFactory(properties, registry, clusterMap, storageManager, accountService);
+        recoveryManager =
+            new RecoveryManager(replicationConfig, clusterMapConfig, storeConfig, storageManager, storeKeyFactory,
+                clusterMap, scheduler, nodeId, networkClientFactory, registry, notificationSystem,
+                storeKeyConverterFactory, serverConfig.serverMessageTransformer, null, null);
+        recoveryManager.start();
+      } else if (serverConfig.serverExecutionMode.equals(ServerExecutionMode.DATA_VERIFICATION_MODE.toString())) {
+        // Backup integrity monitor here because vcr does not have code to store to disk and the code to create that is here
+        RecoveryManager recoveryManager =
+            new RecoveryManager(replicationConfig, clusterMapConfig, storeConfig, storageManager, storeKeyFactory,
+                clusterMap, null, nodeId,
+                new RecoveryNetworkClientFactory(properties, registry, clusterMap, storageManager, accountService),
+                registry, null, storeKeyConverterFactory, serverConfig.serverMessageTransformer,
+                null, null);
+        ReplicationManager serverReplicationManager =
+            new ReplicationManager(replicationConfig, clusterMapConfig, storeConfig, storageManager, storeKeyFactory,
+                clusterMap, null, nodeId,
+                new Http2NetworkClientFactory(new Http2ClientMetrics(registry), new Http2ClientConfig(properties), sslHttp2Factory, time),
+                registry, null, storeKeyConverterFactory, serverConfig.serverMessageTransformer,
+                null, null);
+        backupIntegrityMonitor = new BackupIntegrityMonitor(recoveryManager, serverReplicationManager,
+            (CompositeClusterManager) clusterMap, storageManager, nodeId, findTokenHelper, properties);
+        backupIntegrityMonitor.start();
+      } else if (serverConfig.serverExecutionMode.equals(ServerExecutionMode.DATA_SERVING_MODE.toString())) {
```

Comment:
I also don't quite get some the verification works.

The first step should pick a random partition and use recovery manager to recover data from cloud. Now recovery mananger rely on the clustermap to know which partitions/replicas to recover. In order to make it work, we need to chang clustermap's method so it will return a random partition/replica. And since verification process runs on a different server host, we might not have any replicas in the clustermap for this server host. That indicates we really have to make some change in clustermap.

Reply:
I deal with some of these things in the next pr, since this one is big enough. 
But I agree with your points.

>> The first step should pick a random partition and use recovery manager to recover data from cloud.

Correct. I do this.

>> Now recovery mananger rely on the clustermap to know which partitions/replicas to recover. In order to make it work, we need to chang clustermap's method so it will return a random partition/replica.

Hmm I can see how you got here. RecoveryManager.start() will use a static clustermap to recover selected partitions from cloud. (Or it could be helix clustermap if that's avlb post-disaster, though I have not tested it with helix). 

>> And since verification process runs on a different server host, we might not have any replicas in the clustermap for this server host.

Correct, we do not have any replicas for this host. I work around this in two steps:
1/ Use a static clustermap to discover disks on the verification-host and assign the replica to one of the disks.
2/ Call RecoveryThread::replicate(), instead of RecoveryManager.start()

Reply:
I am fine with the steps, my concern is where those steps are going to happen. 
1. Use a statick clustermap to pick a disk D on the verification-host? Say verification host is HV, will this host placed in the static clustermap? 
2. Replica R should replicate data via RecoveryThread, which means it copies data from cloud. Now do we wait for the copy to finish? or do we verify as the copy goes on?
3. We also created a replication manager to replicate from other peers. I suppose this is trying to get data from on-prem server so we can compare them with data copied from cloud. Where does this data live? Disk D stores data from cloud, do we not store those data on disk and just compare them in memory?

Reply:
>> Use a statick clustermap to pick a disk D on the verification-host? Say verification host is HV, will this host placed in the static clustermap?

Yes, the static clustermap on the host records the hostname, port, disk mountpaths.

>> Replica R should replicate data via RecoveryThread, which means it copies data from cloud. Now do we wait for the copy to finish? or do we verify as the copy goes on?

We wait for recovery to finish. Azure signals us when there is no more data to recover and we can stop the loop.

>> We also created a replication manager to replicate from other peers. I suppose this is trying to get data from on-prem server so we can compare them with data copied from cloud. Where does this data live? Disk D stores data from cloud, do we not store those data on disk and just compare them in memory?

yes, `ReplicationManager` will fetch on-prem data from server hosting replica R. For this, I plan to only fetch metadata + CRC from server and compare it with metadata + CRC from cloud stored in disk D. I do not plan to store metadata + CRC from server on any disk in HV, but just compare and discard or print out mismatches in log/email. Fetching entire blob from server, I think, would be a waste of time and resources and a stress on the server esp when it is busy with customer traffic.